### ISSUE TITLE
Model Object extraction not showing updated value

### ISSUE THREAD

**Issue by ElHombre87** [hishnash/djangochannelsrestframework#16]
**Describe the bug**
I'm not sure if this is an issue with channels, djangochannelsrestframework, or drf.

I'm currently using the @model_observer to watch for model changes via DRF.

This works as intended on all create/update/delete operations. I'm notified of any changes and immediately perform an object extraction.

I figured I'd make it simple and just extract all objects and send it back to the client(s). The clients immediately update the state once the new object is received.

The problem is that the object extracted in the consumer doesn't include the latest update. So I'm sending the unmodified object back to the client(s).

Either I'm incorrectly extracting the model object or there's an async problem, where my object extraction happens to soon.

UPDATE:
Tried extracting the model object with a SyncConsumer, and the result is still the same. It's definitely not an async issue. Perhaps it's because of the DRF serializer? Does that somehow interrupt the ORM processing?

**To Reproduce**
1. Add model observer to Model
2. Subscribe to model change events
3. Extract Model and Print results
4. Make model changes via DRF, which should trigger another model extraction
5. Re-print results

**Expected behavior**
New Model results are the same as the old model results, when it should reflect the recent model changes. 

I would simply like to be notified of all model changes (Which are made via DRF), so I can instantly send the client(s) back the updated model objects.

**LOG**

django==2.2.8
djangorestframework==3.10.3
channels==2.4.0
psycopg2==2.8.3 --no-binary psycopg2

 - OS: [e.g. Ubuntu]
 - Version [e.g. 18.04]

**Additional context**
CODE:

class UserFormConsumer(AsyncAPIConsumer):

    @database_sync_to_async
    def get_userform_object(self, curr_id):
        print('REACHED UserFormConsumer GET USERFORM OBJECT')
        user_form_serializer = serializers.UserFormSerializer(models.UserForm.objects.all(), many=True)
        return user_form_serializer.data

    async def accept(self, subprotocol=None):
        # subscribe to all new changed events
        obj_id = ""
        userform_data = await self.get_userform_object(obj_id)
        print('OLD USERFORM DATA', userform_data)
        await UserFormConsumer.activities_change.subscribe(self)
        await super().accept()

    @model_observer(models.UserForm)  # you Django Model cls goes here.
    async def activities_change(self, message, observer=None, **kwargs):
        # check the message for the action type if you only want to send events for new ones.
        print(observer)
        print('TestingActivitiesChange', message)
        await self.send_json(message)

    async def receive_json(self, content):

        if 'user_id' in content:
            user_id = content['user_id']

            # Get User Group
            current_group_name = 'user_%s' % user_id

            print('USER ID', user_id)
            if content['server_command'] == 'get_all':
                obj_id = ""
                message = content['message']
                userform_data = await self.get_userform_object(obj_id)
                print('Get ALL Message', message)
                print('User Form Data', userform_data)

**Comment by hishnash** [comment]
I belive this might well be related to transactions (currently) this lib picks up changes when the `save` method is called but it does not `queue` them up until an open transaction is committed?  

(just for testing can you try turning of transational requests in django)

**Comment by hishnash** [comment]
in the above code the `receive_json` method is on the same `AsyncAPIConsumer` as your subscribe method?

**Comment by ElHombre87** [comment]
Yes, at first I had everything on one consumer. However when it didn't work, I thought it might be a problem with the websocket connections, so I ended up making another JsonWebsocketConsumer and AsyncJsonWebsocketConsumer that would just be dedicated to extracting and sending the model objects to the clients(s) but both of those consumers didn't work either.

**Comment by hishnash** [comment]
to be clear you what you want to do is when `activities_change` you would like to return the newly updated  results from this method? `get_userform_object`

**Comment by ElHombre87** [comment]
Yup. When objects are modified via DRF REST operations on the client side. I want the activities_change method to notify me and extract the model object via that method, then send the new model object back to the client(s), via groups, where I modify the state.

**Comment by hishnash** [comment]
so the body of `activities_change` is called once for each open consumer (web-socket connection).

If inside this instead of `self.send_json(message)` you use the `id` value in this `message` to query the db `await database_sync_to_async(models.UserForm.objects.get(pk=message.get('pk))`

is the return result of this the updated `UserForm` object

**Comment by ElHombre87** [comment]
I actually haven't tried to access the model object like that because I need all of the objects at once, not just one of them. But I'll give it a try

**Comment by hishnash** [comment]
you could hit  `await self.send_json(await self. get_userform_object())` in that method.

**Comment by ElHombre87** [comment]
> so the body of `activities_change` is called once for each open consumer (web-socket connection).
> 
> If inside this instead of `self.send_json(message)` you use the `id` value in this `message` to query the db `await database_sync_to_async(models.UserForm.objects.get(pk=message.get('pk))`
> 
> is the return result of this the updated `UserForm` object

The result still isn't the updated version after just accessing the individual object.

Is this strange behavior? I'm surprised no one else is having this issue.

**Comment by hishnash** [comment]
> "await self.send_json(await self. get_userform_object()) in that method."

No it will just let you get all of them at once.

> Is this strange behavior? I'm surprised no one else is having this issue.

Yes, unless you are within a transition. Can you share how you are updating the `UserForm`?

**Comment by ElHombre87** [comment]
The Model objects are updated via DRF POST/PUT requests on the client side. I'm using Quasar (Vue.js framework) for the frontend.

When the user logs in, they subscribe to the consumer like so:

```
            Vue.prototype.$liveSocket = new WebSocket(`${window.location.protocol === 'https:' ? 'wss' : 'ws'}://localhost:8000/userform/live/`)
            Vue.prototype.$liveSocket.onmessage = function (m) {
                Vue.prototype.$liveSocket.send(JSON.stringify({
                    'server_command': 'get_all',
                    'message': m,
                    'user_id': state.user_id
                }));
            }
```

**Comment by hishnash** [comment]
hmm I will take a look at this later today. seems strange.

**Comment by hishnash** [comment]
In your above JS code if you wait 1 second before querying all the items does it help?

```js
setTimeout(
    Vue.prototype.$liveSocket.send(JSON.stringify({
                    'server_command': 'get_all',
                    'message': m,
                    'user_id': state.user_id
})), 1000)
```

**Comment by ElHombre87** [comment]
You, sir, are the man.

Nice one, it worked.

**Comment by hishnash** [comment]
hmmm ... I still dont feel this is solved. If your sure your normal Django rest framework views are not in a translation im confused how this could happen. (do you by any chance have some other reason for it to take time for changes to the db to replicate)? eg do you have a readonly db slave set up?

The `@model_observer` listens to `django's` `post_save` event that is fired after the save hits the db.

**Comment by ElHombre87** [comment]
It's very strange, I haven't seen waits applied in any documentation, but it definitely works. I removed it and it stopped working, then added a half-second wait and it works again.

```
                setTimeout(() => {
                    Vue.prototype.$liveSocket.send(JSON.stringify({
                        'server_command': 'get_all',
                        'message': m,
                        'user_id': state.user_id
                    }));
                }, 500);
```
There's nothing special about my POSTGRES database setup or my Views.

```
class UserFormViewSet(viewsets.ModelViewSet):
    queryset = UserForm.objects.all()
    serializer_class = serializers.UserFormSerializer
    permission_classes = (permissions.IsAuthenticated,)
    lookup_field = 'slug'

    def get_queryset(self):
        queryset = UserForm.objects.all()
        return queryset
```

**Comment by hishnash** [comment]
@ElHombre87 i have fixed this in version `0.0.5` you will need to turn on (`ATOMIC_REQUESTS`)[https://docs.djangoproject.com/en/3.0/ref/settings/#std:setting-DATABASE-ATOMIC_REQUESTS] or wrapp you handler with the `atomic` decorator.
